compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -2094,21 +2194,15 @@ object Parsers {
      */
     def classConstr(owner: Name, isCaseClass: Boolean = false): DefDef = atPos(in.lastOffset) {
       val tparams = typeParamClauseOpt(ParamOwner.Class)
-      val cmods = constrModsOpt(owner)
+      val cmods = fromWithinClassConstr(constrModsOpt(owner))
       val vparamss = paramClauses(owner, isCaseClass)
       makeConstructor(tparams, vparamss).withMods(cmods)
     }
 
-    /** ConstrMods        ::=  AccessModifier
-     *                      |  Annotation {Annotation} (AccessModifier | `this')
+    /** ConstrMods        ::=  {Annotation} [AccessModifier]
      */
-    def constrModsOpt(owner: Name): Modifiers = {
-      val mods = modifiers(accessModifierTokens, annotsAsMods())
-      if (mods.hasAnnotations && !mods.hasFlags)
-        if (in.token == THIS) in.nextToken()
-        else syntaxError(AnnotatedPrimaryConstructorRequiresModifierOrThis(owner), mods.annotations.last.pos)
```

Comment:
This means we can drop this `Message` class (but we must be careful to leave a dummy id in its place in the `ErrorMessageID` enum to avoid changing the error number of every subsequent message).